### DISCUSSION TITLE
Set Guicursor to nil or 0%

### DISCUSSION THREAD

**Discussion by Vikareaj** [neovim/neovim#24612] [Q&A] [answered]
### Problem

  Greetings,

  I use nvim for writng prose. As such, nvim's initial state is unusable for my particular needs and I have heavily (and awkwardly,) customized it (in a way that would perhaps baffle programmers.)

  Still, nvim's chimeraeness  has allowed me to create a distraction-free typing environment I have come to love and consider superior to the (many) other editors that are traditionally thought of as prose-centric.

  A few days ago I decided to tinker a bit in my vimrc and fix the one last thing that I see as a distraction.

  The guicursor in insert mode.

  My plan was to remove it completely (when in  insert mode,) and emulate the typing experience of a typewriter.

  To that effect I set guicursor to i:ver0 and i:nil. The first I inferred from the wiki (neovim.io) while the latter was just a stab in the dark.

  i:ver0 returns "invalid percentage," and i:nil also outputs an error message. Neither works.

  My programming skills are lackluster to say the least, so I apologize in advance if nvim is already capable of the feature I am requesting and I just missed the how-to in the wiki.

  Thank you for your time your efforts and nvim in general, by a non-conventional user of nvim,
  
  Anthony

### Expected behavior

Be able to set guicursor to nil / 0% in insert mode for truly distraction-free environment.

**Comment by justinmk** [comment]
Depends on the UI (or your terminal).

This is somewhat possible already by sending a `busy_start` event from the UI, see `:help ui-global`.

```
["busy_start"]  
["busy_stop"]  
	Indicates to the UI that it must stop rendering the cursor. This event
	is misnamed and does not actually have anything to do with busyness.
```

**Reply by Vikareaj** [reply]
** Apologies, I just realized my previous responce was a new comment on the thread instead of a responce.

Thank you very much for the swift responce!

Could you provide an example use? My mostly poor command-line skills are only producing a string of "Aborted (core dumped)"

Also, a question: Will busy_start, busy_stop work for only one mode (the insert mode,) or will it apply universally?

Mu OS is linux and my terminal kitty. (Although I am not married to the terminal if it is the one causing issues.)

Again, thank you.

Ps: Following the wiki as best I could I tried setting vim.opt.busy_start and vim.opt.busy in my vimrc (settings.lua) as other items like showtabline seem to be set this way, but of course it doesn't work.

**Comment by Vikareaj** [comment]
Thank you very much for the swift responce!

  Could you provide an example use? My mostly poor command-line skills are only producing a string of "Aborted (core dumped)" 

  Also, a question: Will busy_start, busy_stop work for only one mode (the insert mode,) or will it apply universally?

 Mu OS is linux and my terminal kitty. (Although I am not married to the terminal if it is the one causing issues.)

Again, thank you.

**Reply by ad-chaos** [reply]
I am not sure what you are seeking for exactly, does `:set guicursor=` solve your problem?

**Reply by Vikareaj** [reply]
First of all, thanks for replying.

No, guicursor does not (cannot) set the behaviour I am looking for. (I want *no* cursor. Not a vertical or a block one, just for the cursor to not be rendered.)

The option that governs the behaviour I am looking for is "busy_start" and "busy_stop" as explained by justinmk.

But I don't know how to set it.

Guicursor I can set, vim.opt.guicursor = "..." works just fine.

But I can't, for the love of gods, find how/where to set "busy_start" and "busy_stop"

**Reply by ad-chaos** [reply, marked as answer]
Ye neovim doesn't expose it, you can do the following, but expect it to break in future versions
```lua
-- typewriter.lua

local ffi = require("ffi")

ffi.cdef([[
void ui_busy_start(void);
void ui_busy_stop(void);
]])

vim.api.nvim_create_autocmd({"InsertEnter"}, {
  callback = function()
    ffi.C.ui_busy_start()
  end
})

vim.api.nvim_create_autocmd({"InsertLeave"}, {
  callback = function()
    ffi.C.ui_busy_stop()
  end
})
```

Test it out by doing `nvim --clean -u typewriter.lua`

**Reply by Vikareaj** [reply]
It works!

Can't thank you enough. I would have never come up with those commands on my own.

Here's hoping when it does break, a simpler way to accomplish it is introduced.

Thanks a million, again.
